src/routeSpec/RouteSpec.js

```diff
@@ -0,0 +1,52 @@
+'use strict'
+const QueryParameter = require('./QueryParameter')
+const _VALID_METHODS = ['CONNECT', 'DELETE', 'GET', 'HEAD', 'OPTIONS', 'PATCH', 'POST', 'PUT', 'TRACE']
+const _URI_SEGMENT_REGEX = '([a-zA-Z0-9\\$\\-_\\@\\.&\\+\\-\\!*"\'\\(\\)\\,](\\%[0-9a-fA-F]{2})?)+'
+const _PATH_PARAM_REGEX = `^${_URI_SEGMENT_REGEX}$`
+const _PATH_REGEX = `^\\/${_URI_SEGMENT_REGEX}$`
+const isValidPathDefinition = paths => paths.every(path => path.match(_PATH_REGEX))
+const isValidMethodDefinition = methods => methods.every(method => _VALID_METHODS.includes(method.toUpperCase()))
+const isValidPathParametersDefinition = pathParameters => pathParameters.every(param => param.match(_PATH_PARAM_REGEX))
+const isValidQueryParametersDefinition = queryParameters => queryParameters.every(param => param instanceof QueryParameter)
+const getPaths = paths => isValidPathDefinition(paths) ? paths : false
+const getMethods = methods => isValidMethodDefinition(methods) ? methods.map(method => method.toUpperCase()) : false
+const getPathParameters = pathParameters => isValidPathParametersDefinition(pathParameters) ? pathParameters : false
+const getQueryParameters = queryParameters => isValidQueryParametersDefinition(queryParameters) ? queryParameters : false
+
+class RouteSpec {
+  /**
+   * Defines a route and the parameters that can be used with the route.
+   * @param {Array.<String>} paths
+   * @param {Array.<String>} methods
+   * @param {Array.<String>} pathParameters
+   * @param {Array.<QueryParameter>} queryParameters
+   */
+  constructor (paths, methods, pathParameters, queryParameters) {
+    this._paths = getPaths(paths) || (() => { throw new Error('Invalid paths definition') })()
+    this._methods = getMethods(methods) || (() => { throw new Error('Invalid methods definition') })()
+    this._pathParameters = getPathParameters(pathParameters || []) || (() => { throw new Error('Bad path parameters definition') })()
```

Comment:
here we are implicitly saying that empty pathParameters s valid. So this logic should go to `allMethodsAreValid`